In [1]:
import numpy as np
import pandas as pd
import plotly.graph_objs as go
import plotly.express as px
from plotly.subplots import make_subplots
from scipy.stats import norm

In [2]:
# Для нейронок
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import Dataset, DataLoader

In [3]:
# Arch/Garch и другие модели
from arch import arch_model

In [4]:
df = pd.read_json("./data.json")
df.columns = ["BOARDID", "SECID", "TRADEDATE", "SHORTNAME", "NAME", "CLOSE", "OPEN", "HIGH", "LOW", "VALUE", "DURATION", "YIELD", "DECIMALS", "CAPITALIZATION", "CURRENCYID", "DIVISOR", "TRADINGSESSION", "VOLUME", "TRADE_SESSION_DATE", "RECALC_DATE"]
df

,BOARDID,SECID,TRADEDATE,SHORTNAME,NAME,CLOSE,OPEN,HIGH,LOW,VALUE,DURATION,YIELD,DECIMALS,CAPITALIZATION,CURRENCYID,DIVISOR,TRADINGSESSION,VOLUME,TRADE_SESSION_DATE,RECALC_DATE
0,SNDX,IMOEX,2010-01-12,Индекс МосБиржи,Индекс МосБиржи,1427.67,1444.78,1445.70,1424.73,3.882154e+10,NaN,NaN,2,NaN,RUB,NaN,3,NaN,2010-01-12,NaN
1,SNDX,IMOEX,2010-01-13,Индекс МосБиржи,Индекс МосБиржи,1435.01,1427.62,1444.86,1412.87,4.222875e+10,NaN,NaN,2,NaN,RUB,NaN,3,NaN,2010-01-13,NaN
2,SNDX,IMOEX,2010-01-14,Индекс МосБиржи,Индекс МосБиржи,1455.65,1435.01,1456.34,1435.01,4.451159e+10,NaN,NaN,2,NaN,RUB,NaN,3,NaN,2010-01-14,NaN
3,SNDX,IMOEX,2010-01-15,Индекс МосБиржи,Индекс МосБиржи,1452.67,1455.47,1469.55,1448.68,4.267554e+10,NaN,NaN,2,NaN,RUB,NaN,3,NaN,2010-01-15,NaN
4,SNDX,IMOEX,2010-01-18,Индекс МосБиржи,Индекс МосБиржи,1479.53,1452.67,1479.76,1444.90,4.304998e+10,NaN,NaN,2,NaN,RUB,NaN,3,NaN,2010-01-18,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4072,SNDX,IMOEX,2026-03-30,Индекс МосБиржи,Индекс МосБиржи,2800.69,2796.46,2814.54,2775.80,6.808478e+10,0.0,0.0,2,7.374585e+12,RUB,2.633132e+09,3,NaN,2026-03-30,NaN
4073,SNDX,IMOEX,2026-03-31,Индекс МосБиржи,Индекс МосБиржи,2776.37,2801.02,2801.10,2766.28,5.980559e+10,0.0,0.0,2,7.310538e+12,RUB,2.633132e+09,3,NaN,2026-03-31,NaN
4074,SNDX,IMOEX,2026-04-01,Индекс МосБиржи,Индекс МосБиржи,2775.24,2770.81,2787.45,2764.84,5.491598e+10,0.0,0.0,2,7.307560e+12,RUB,2.633132e+09,3,NaN,2026-04-01,NaN
4075,SNDX,IMOEX,2026-04-02,Индекс МосБиржи,Индекс МосБиржи,2774.41,2787.52,2787.68,2767.99,5.170786e+10,0.0,0.0,2,7.305392e+12,RUB,2.633132e+09,3,NaN,2026-04-02,NaN


In [59]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=df["TRADEDATE"],
    y=df["CLOSE"],
    mode='lines',
    name='IMOEX', 
    line_shape='hv'
))

fig.update_xaxes(title_text='день - x')
fig.update_yaxes(title_text='индекс - y')
fig.update_layout(title='IMOEX')
fig.show()

fig.write_html(
    "./plots/interactive_plot_IMOEX_2010_2026.html",
    auto_open=False,
    full_html=True,
    config={
        'scrollZoom': True, 
        'displayModeBar': True, 
        'displaylogo': False, 
        'modeBarButtonsToRemove': ['lasso2d', 'select2d']
    }
)

In [6]:
df['LogReturn'] = np.log(df['CLOSE'] / df['CLOSE'].shift(1))
returns = df['LogReturn'].dropna() * 100

In [58]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=df["TRADEDATE"],
    y=returns,
    mode='lines',
    name='IMOEX', 
    line_shape='hv'
))

fig.update_xaxes(title_text='день - x')
fig.update_yaxes(title_text='индекс - y')
fig.update_layout(title='IMOEX')
fig.show()

fig.write_html(
    "./plots/interactive_log_IMOEX.html",
    auto_open=False,
    full_html=True,
    config={
        'scrollZoom': True, 
        'displayModeBar': True, 
        'displaylogo': False, 
        'modeBarButtonsToRemove': ['lasso2d', 'select2d']
    }
)

In [8]:
model=arch_model(returns[:2750], vol='Garch', p=1, o=0, q=1, dist='Normal')
results=model.fit()
print(results.summary())
print(results.params)

Iteration:      1,   Func. Count:      6,   Neg. LLF: 7375903284095.698
Iteration:      2,   Func. Count:     15,   Neg. LLF: 4779608194.721778
Iteration:      3,   Func. Count:     23,   Neg. LLF: 5834.617792172406
Iteration:      4,   Func. Count:     30,   Neg. LLF: 4557.627451188812
Iteration:      5,   Func. Count:     38,   Neg. LLF: 4317.064406093726
Iteration:      6,   Func. Count:     44,   Neg. LLF: 4290.9832080248325
Iteration:      7,   Func. Count:     49,   Neg. LLF: 4290.95660676393
Iteration:      8,   Func. Count:     54,   Neg. LLF: 4290.954490953034
Iteration:      9,   Func. Count:     59,   Neg. LLF: 4290.953758433864
Iteration:     10,   Func. Count:     64,   Neg. LLF: 4290.953757171089
Iteration:     11,   Func. Count:     68,   Neg. LLF: 4290.953757171241
Optimization terminated successfully    (Exit mode 0)
            Current function value: 4290.953757171089
            Iterations: 11
            Function evaluations: 68
            Gradient evaluations: 11

In [9]:

window = 90  

predictions = []
dates = []

for t in range(2750, len(returns)):
    train_data = returns[t - window : t]
    
    model = arch_model(train_data, vol='Garch', p=1, q=1, dist='Normal')
    result = model.fit(disp='off')  
    
    forecast = result.forecast(horizon=1)
    predicted_var = forecast.variance.values[-1, 0]  # σ²
    
    predictions.append(predicted_var)
    dates.append(df['TRADEDATE'].iloc[t])

predictions = np.array(predictions)
predicted_vol = np.sqrt(predictions)  # σ (стд. отклонение)

In [10]:
# Реализованная волатильность - скользящее σ за 20 дней
realized_vol = returns.rolling(window=20).std()

# Берём те же индексы что и предсказания
realized_test = realized_vol.iloc[2750:].values

In [11]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

mse = mean_squared_error(realized_test, predicted_vol)
mae = mean_absolute_error(realized_test, predicted_vol)
r2 = r2_score(realized_test, predicted_vol)

print(f"GARCH: MSE={mse:.6f}, MAE={mae:.6f}, R2={r2:.4f}")

GARCH: MSE=1.065735, MAE=0.345033, R2=0.3787


In [12]:
fig = go.Figure()
fig.add_trace(go.Scatter(x=dates, y=realized_test, name='Реальность'))
fig.add_trace(go.Scatter(x=dates, y=predicted_vol, name='GARCH'))
fig.update_layout(title='Волатильность: GARCH vs реальность')
fig.show()

## Подготовка данных для нейронок

In [13]:
train_mask = df['TRADEDATE'] < '2020-01-01'
val_mask = (df['TRADEDATE'] >= '2020-01-01') & (df['TRADEDATE'] < '2022-01-01')
test_mask = df['TRADEDATE'] >= '2022-01-01'

print(f"Train: {train_mask.sum()} дней (2010–2019)")
print(f"Val:   {val_mask.sum()} дней (2020–2021)")
print(f"Test:  {test_mask.sum()} дней (2022–2026)")

Train: 2510 дней (2010–2019)
Val:   505 дней (2020–2021)
Test:  1062 дней (2022–2026)


In [31]:
# Реализованная волатильность
# returns: df['LogReturn'] * 100

# Ground truth variance как в статье GINN:
# σ²_t = (r_t − μ̂_t)²
# где μ̂_t — скользящее среднее за 90 дней (AR-оценка)

rolling_mean = returns.rolling(window=90).mean()
raw_var = (returns - rolling_mean) ** 2

ground_truth_var = raw_var.rolling(window=5).mean()

valid_idx = ground_truth_var.dropna().index
print(f"Валидных точек после rolling: {len(valid_idx)}")

Валидных точек после rolling: 3983


In [32]:
# Формируем окна для нейронок
WINDOW = 90 

gt_var = ground_truth_var.dropna().values  # ground truth variance
gt_var = np.log1p(gt_var) # Волатильность имеет тяжёлый правый хвост (бывают редкие огромные значения). Log сжимает диапазон
trade_dates = df.loc[ground_truth_var.dropna().index, 'TRADEDATE'].values

X_windows = []
y_targets = []
window_dates = []

for i in range(WINDOW, len(gt_var)):
    X_windows.append(gt_var[i - WINDOW : i])  # 90 дней σ²
    y_targets.append(gt_var[i])                # σ² на следующий день
    window_dates.append(trade_dates[i])

X_windows = np.array(X_windows)  # shape: (N, 90)
y_targets = np.array(y_targets)  # shape: (N,)
window_dates = np.array(window_dates)

print(f"X shape: {X_windows.shape}")
print(f"y shape: {y_targets.shape}")
print(f"Период: {window_dates[0]} - {window_dates[-1]}")

X shape: (3893, 90)
y shape: (3893,)
Период: 2010-10-04 - 2026-04-03


In [33]:
window_dates_dt = window_dates.astype('datetime64')

train_idx = window_dates_dt < np.datetime64('2020-01-01')
val_idx = (window_dates_dt >= np.datetime64('2020-01-01')) & (window_dates_dt < np.datetime64('2022-01-01'))
test_idx = window_dates_dt >= np.datetime64('2022-01-01')

X_train, y_train = X_windows[train_idx], y_targets[train_idx]
X_val, y_val = X_windows[val_idx], y_targets[val_idx]
X_test, y_test = X_windows[test_idx], y_targets[test_idx]
dates_test = window_dates[test_idx]  # оригинальные строки, если нужны

print(f"Train: {X_train.shape[0]} окон")
print(f"Val:   {X_val.shape[0]} окон")
print(f"Test:  {X_test.shape[0]} окон")

Train: 2326 окон
Val:   505 окон
Test:  1062 окон


In [34]:
# === Нормализация ===
train_mean = X_train.mean()
train_std = X_train.std()

X_train_norm = (X_train - train_mean) / train_std
X_val_norm = (X_val - train_mean) / train_std
X_test_norm = (X_test - train_mean) / train_std

y_train_norm = (y_train - train_mean) / train_std
y_val_norm = (y_val - train_mean) / train_std
y_test_norm = (y_test - train_mean) / train_std

print(f"Train mean: {train_mean:.6f}, std: {train_std:.6f}")

Train mean: 0.717934, std: 0.481353


In [35]:
class VolatilityDataset(torch.utils.data.Dataset):
    def __init__(self, X, y):
        # X shape: (N, 90) -> (N, 90, 1) для LSTM
        self.X = torch.FloatTensor(X).unsqueeze(-1)  # (N, 90, 1)
        self.y = torch.FloatTensor(y)                 # (N,)
    
    def __len__(self):
        return len(self.y)
    
    def __getitem__(self, idx):
        return self.X[idx], self.y[idx]

train_dataset = VolatilityDataset(X_train_norm, y_train_norm)
val_dataset = VolatilityDataset(X_val_norm, y_val_norm)
test_dataset = VolatilityDataset(X_test_norm, y_test_norm)

BATCH_SIZE = 64

train_loader = DataLoader(train_dataset, batch_size=BATCH_SIZE, shuffle=True)
val_loader = DataLoader(val_dataset, batch_size=BATCH_SIZE, shuffle=False)
test_loader = DataLoader(test_dataset, batch_size=BATCH_SIZE, shuffle=False)

print(f"Train batches: {len(train_loader)}")
print(f"Val batches:   {len(val_loader)}")
print(f"Test batches:  {len(test_loader)}")

# Проверка одного батча
X_batch, y_batch = next(iter(train_loader))
print(f"\nBatch X shape: {X_batch.shape}")   # (64, 90, 1)
print(f"Batch y shape: {y_batch.shape}")     # (64,)

Train batches: 37
Val batches:   8
Test batches:  17

Batch X shape: torch.Size([64, 90, 1])
Batch y shape: torch.Size([64])


### Создание и обучение LSTM

In [36]:
class VolatilityLSTM(nn.Module):
    """
    LSTM модель для предсказания волатильности
    Архитектура из статьи:
    - 3 LSTM слоя с шириной 256
    - Dropout слои между LSTM слоями
    - 2 линейных слоя (fully connected)
    - 1 BatchNorm слой
    - 1 ReLU активация
    """
    def __init__(self, input_size=1, hidden_size=256, num_layers=3, dropout=0.2):
        super(VolatilityLSTM, self).__init__()
        
        self.hidden_size = hidden_size
        self.num_layers = num_layers
        
        self.lstm = nn.LSTM(
            input_size=input_size,
            hidden_size=hidden_size,
            num_layers=num_layers,
            dropout=dropout if num_layers > 1 else 0,
            batch_first=True
        )
        
        self.batch_norm = nn.BatchNorm1d(hidden_size)
        
        self.fc1 = nn.Linear(hidden_size, 128)
        self.fc2 = nn.Linear(128, 1)
        
        self.dropout = nn.Dropout(dropout)
        
        self.relu = nn.ReLU()
        
    def forward(self, x):
        # x shape: (batch_size, sequence_length, input_size)
        
        lstm_out, (hidden, cell) = self.lstm(x)
        # lstm_out: (batch_size, sequence_length, hidden_size)
        
        last_output = lstm_out[:, -1, :]  # (batch_size, hidden_size)
        normalized = self.batch_norm(last_output)
        
        fc1_out = self.relu(self.fc1(normalized))
        fc1_out = self.dropout(fc1_out)
        
        output = self.fc2(fc1_out)
        
        return output.squeeze(-1)

In [37]:
LSTM_EPOCHS = 300
lstm_model = VolatilityLSTM()
criterion = nn.MSELoss()
optimizer = optim.AdamW(lstm_model.parameters(), lr=1e-3)
scheduler = optim.lr_scheduler.ReduceLROnPlateau(optimizer, mode='min', patience=20, factor=0.5)

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
lstm_model.to(device)


VolatilityLSTM(
  (lstm): LSTM(1, 256, num_layers=3, batch_first=True, dropout=0.2)
  (batch_norm): BatchNorm1d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (fc1): Linear(in_features=256, out_features=128, bias=True)
  (fc2): Linear(in_features=128, out_features=1, bias=True)
  (dropout): Dropout(p=0.2, inplace=False)
  (relu): ReLU()
)

In [38]:
best_val_loss = float('inf')
best_model_state = None
patience_counter = 0
PATIENCE = 30

for epoch in range(LSTM_EPOCHS):
    lstm_model.train()
    train_losses = []
    
    for X_batch, y_batch in train_loader:
        optimizer.zero_grad()
        outputs = lstm_model(X_batch.to(device))
        loss = criterion(outputs, y_batch.to(device))
        loss.backward()
        torch.nn.utils.clip_grad_norm_(lstm_model.parameters(), max_norm=1.0)
        optimizer.step()
        
        train_losses.append(loss.item())
    
    avg_train_loss = np.mean(train_losses)
    
    lstm_model.eval()
    val_losses = []
    
    with torch.no_grad():
        for X_batch, y_batch in val_loader:
            outputs = lstm_model(X_batch.to(device))
            loss = criterion(outputs, y_batch.to(device))
            val_losses.append(loss.item())
    
    avg_val_loss = np.mean(val_losses)
    scheduler.step(avg_val_loss)

    if avg_val_loss < best_val_loss:
        best_val_loss = avg_val_loss
        best_model_state = lstm_model.state_dict().copy()
        patience_counter = 0
    else:
        patience_counter += 1
    
    if patience_counter >= PATIENCE:
        print(f"Ранняя остановка на эпохе {epoch+1}")
        break

    if (epoch + 1) % 25 == 0:
        current_lr = optimizer.param_groups[0]['lr']
        print(f"Измененный LR: {current_lr:.6f}")
    
    if (epoch + 1) % 10 == 0 or epoch == 0:
        print(f"Эпоха {epoch+1}/{LSTM_EPOCHS} - Train Loss: {avg_train_loss:.6f}, Val Loss: {avg_val_loss:.6f}")


lstm_model.load_state_dict(best_model_state)

Эпоха 1/300 - Train Loss: 0.473773, Val Loss: 1.455231
Эпоха 10/300 - Train Loss: 0.224809, Val Loss: 0.233623
Эпоха 20/300 - Train Loss: 0.210969, Val Loss: 0.235957
Измененный LR: 0.001000
Эпоха 30/300 - Train Loss: 0.202324, Val Loss: 0.281529
Эпоха 40/300 - Train Loss: 0.181173, Val Loss: 0.210178
Измененный LR: 0.001000
Эпоха 50/300 - Train Loss: 0.177165, Val Loss: 0.176493
Эпоха 60/300 - Train Loss: 0.166196, Val Loss: 0.173006
Эпоха 70/300 - Train Loss: 0.155940, Val Loss: 0.167299
Измененный LR: 0.000500
Эпоха 80/300 - Train Loss: 0.147222, Val Loss: 0.202858
Эпоха 90/300 - Train Loss: 0.144653, Val Loss: 0.170534
Ранняя остановка на эпохе 96


<All keys matched successfully>

In [39]:
lstm_model.eval()
all_preds = []
all_targets = []

with torch.no_grad():
    for X_batch, y_batch in test_loader:
        preds = lstm_model(X_batch.to(device)).cpu().numpy()
        all_preds.extend(preds)
        all_targets.extend(y_batch.numpy())

all_preds = np.array(all_preds)
all_targets = np.array(all_targets)

# Денормализация
preds_denorm = np.expm1(all_preds * train_std + train_mean)
targets_denorm = np.expm1(all_targets * train_std + train_mean)

mse = mean_squared_error(targets_denorm, preds_denorm)
mae = mean_absolute_error(targets_denorm, preds_denorm)
r2 = r2_score(targets_denorm, preds_denorm)

print(f"LSTM Test: MSE={mse:.4f}, MAE={mae:.4f}, R²={r2:.4f}")

LSTM Test: MSE=350.9200, MAE=1.9464, R²=0.5188


In [44]:
# Используем returns

returns_array = returns.values
dates_array = df['TRADEDATE'].values[1:]  

# Индексы test окон в исходных returns
test_positions = []
for d in dates_test:
    idx = np.where(dates_array == d)[0]
    if len(idx) > 0:
        test_positions.append(idx[0])

garch_preds_raw = []
for pos in test_positions:
    train_data = returns.iloc[pos - 90 : pos]
    try:
        m = arch_model(train_data, vol='Garch', p=1, q=1, dist='Normal')
        r = m.fit(disp='off')
        f = r.forecast(horizon=1)
        garch_preds_raw.append(f.variance.values[-1, 0])
    except:
        garch_preds_raw.append(np.nan)

garch_preds_raw = np.array(garch_preds_raw)

In [45]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=dates_test, y=targets_denorm,
    name='Реальность', opacity=0.6
))

fig.add_trace(go.Scatter(
    x=dates_test, y=preds_denorm,
    name='LSTM', opacity=0.8
))

fig.add_trace(go.Scatter(
    x=dates_test, y=garch_preds_raw,
    name='GARCH', opacity=0.8
))

fig.update_layout(
    title='Волатильность: LSTM vs GARCH vs реальность',
    xaxis_title='Дата',
    yaxis_title='σ²'
)
fig.show()

In [48]:
mask = ~np.isnan(garch_preds_raw)
garch_mse = mean_squared_error(targets_denorm[mask], garch_preds_raw[mask])
garch_mae = mean_absolute_error(targets_denorm[mask], garch_preds_raw[mask])
garch_r2 = r2_score(targets_denorm[mask], garch_preds_raw[mask])

print(f"{'Модель':<12} {'MSE':>10} {'MAE':>10} {'R²':>10}")
print(f"{'GARCH':<12} {garch_mse:>10.4f} {garch_mae:>10.4f} {garch_r2:>10.4f}")
print(f"{'LSTM':<12} {mse:>10.4f} {mae:>10.4f} {r2:>10.4f}")

print(f"LSTM targets: min={targets_denorm.min():.2f}, max={targets_denorm.max():.2f}, mean={targets_denorm.mean():.2f}")
print(f"GARCH preds:  min={np.nanmin(garch_preds_raw):.2f}, max={np.nanmax(garch_preds_raw):.2f}, mean={np.nanmean(garch_preds_raw):.2f}")
print(f"LSTM preds:   min={preds_denorm.min():.2f}, max={preds_denorm.max():.2f}, mean={preds_denorm.mean():.2f}")

Модель              MSE        MAE         R²
GARCH          261.2266     2.3915     0.6418
LSTM           350.9200     1.9464     0.5188
LSTM targets: min=0.02, max=414.27, mean=4.08
GARCH preds:  min=0.12, max=731.25, mean=4.32
LSTM preds:   min=0.26, max=207.31, mean=2.87


## модель Transformer

In [49]:
class VolatilityTransformer(nn.Module):
    def __init__(self, input_size=1, d_model=64, nhead=4, num_layers=2, dropout=0.2, seq_len=90):
        super().__init__()
        
        self.embedding = nn.Linear(input_size, d_model)
        self.pos_encoding = nn.Parameter(torch.randn(1, seq_len, d_model) * 0.01)
        
        encoder_layer = nn.TransformerEncoderLayer(
            d_model=d_model,
            nhead=nhead,
            dim_feedforward=d_model * 4,
            dropout=dropout,
            batch_first=True
        )
        self.transformer = nn.TransformerEncoder(encoder_layer, num_layers=num_layers)
        
        self.head = nn.Sequential(
            nn.Linear(d_model, d_model),
            nn.ReLU(),
            nn.Dropout(dropout),
            nn.Linear(d_model, 1)
        )
    
    def forward(self, x):
        # x: (batch, 90, 1)
        x = self.embedding(x) + self.pos_encoding  # (batch, 90, d_model)
        x = self.transformer(x)                      # (batch, 90, d_model)
        x = x[:, -1, :]                              # берём последний timestep
        return self.head(x).squeeze(-1)               # (batch,)

In [50]:
TRANS_EPOCHS = 300
trans_model = VolatilityTransformer()
trans_criterion = nn.MSELoss()
trans_optimizer = optim.AdamW(trans_model.parameters(), lr=1e-3)
trans_scheduler = optim.lr_scheduler.ReduceLROnPlateau(trans_optimizer, mode='min', patience=20, factor=0.5)

trans_model.to(device)

best_val_loss = float('inf')
best_model_state = None
patience_counter = 0
PATIENCE = 30

for epoch in range(TRANS_EPOCHS):
    trans_model.train()
    train_losses = []
    
    for X_batch, y_batch in train_loader:
        trans_optimizer.zero_grad()
        outputs = trans_model(X_batch.to(device))
        loss = trans_criterion(outputs, y_batch.to(device))
        loss.backward()
        torch.nn.utils.clip_grad_norm_(trans_model.parameters(), max_norm=1.0)
        trans_optimizer.step()
        train_losses.append(loss.item())
    
    avg_train_loss = np.mean(train_losses)
    
    trans_model.eval()
    val_losses = []
    with torch.no_grad():
        for X_batch, y_batch in val_loader:
            preds = trans_model(X_batch.to(device))
            loss = trans_criterion(preds, y_batch.to(device))
            val_losses.append(loss.item())
    
    avg_val_loss = np.mean(val_losses)
    trans_scheduler.step(avg_val_loss)
    
    if avg_val_loss < best_val_loss:
        best_val_loss = avg_val_loss
        best_model_state = {k: v.clone() for k, v in trans_model.state_dict().items()}
        patience_counter = 0
    else:
        patience_counter += 1
    
    if (epoch + 1) % 10 == 0:
        print(f"Эпоха {epoch+1}/{TRANS_EPOCHS} - Train: {avg_train_loss:.6f}, Val: {avg_val_loss:.6f}")
    
    if patience_counter >= PATIENCE:
        print(f"Ранняя остановка на эпохе {epoch+1}")
        break

trans_model.load_state_dict(best_model_state)

Эпоха 10/300 - Train: 0.232714, Val: 0.215630
Эпоха 20/300 - Train: 0.214851, Val: 0.265197
Эпоха 30/300 - Train: 0.197424, Val: 0.238643
Эпоха 40/300 - Train: 0.179376, Val: 0.195245
Эпоха 50/300 - Train: 0.167477, Val: 0.224336
Эпоха 60/300 - Train: 0.167607, Val: 0.241327
Эпоха 70/300 - Train: 0.155111, Val: 0.222007
Ранняя остановка на эпохе 78


<All keys matched successfully>

In [52]:
trans_model.eval()
trans_preds = []

with torch.no_grad():
    for X_batch, y_batch in test_loader:
        preds = trans_model(X_batch.to(device)).cpu().numpy()
        trans_preds.extend(preds)

trans_preds = np.array(trans_preds)
trans_preds_denorm = np.expm1(trans_preds * train_std + train_mean)

trans_mse = mean_squared_error(targets_denorm, trans_preds_denorm)
trans_mae = mean_absolute_error(targets_denorm, trans_preds_denorm)
trans_r2 = r2_score(targets_denorm, trans_preds_denorm)

print(f"{'Модель':<15} {'MSE':>10} {'MAE':>10} {'R²':>10}")
print(f"{'GARCH':<15} {garch_mse:>10.4f} {garch_mae:>10.4f} {garch_r2:>10.4f}")
print(f"{'LSTM':<15} {mse:>10.4f} {mae:>10.4f} {r2:>10.4f}")
print(f"{'Transformer':<15} {trans_mse:>10.4f} {trans_mae:>10.4f} {trans_r2:>10.4f}")

Модель                 MSE        MAE         R²
GARCH             261.2266     2.3915     0.6418
LSTM              350.9200     1.9464     0.5188
Transformer       635.6202     2.4518     0.1283


In [60]:
fig = go.Figure()
fig.add_trace(go.Scatter(x=dates_test, y=targets_denorm, name='Реальность', opacity=0.5))
fig.add_trace(go.Scatter(x=dates_test, y=garch_preds_raw, name='GARCH', opacity=0.7))
fig.add_trace(go.Scatter(x=dates_test, y=preds_denorm, name='LSTM', opacity=0.7))
fig.add_trace(go.Scatter(x=dates_test, y=trans_preds_denorm, name='Transformer', opacity=0.7))
fig.update_layout(title='Волатильность: все модели vs реальность',
                  xaxis_title='Дата', yaxis_title='σ²')
fig.show()

fig.write_html(
    "./plots/interactive_plot_GARCH_LSTM_Transformer_TEST_2022_2026.html",
    auto_open=False,
    full_html=True,
    config={
        'scrollZoom': True, 
        'displayModeBar': True, 
        'displaylogo': False, 
        'modeBarButtonsToRemove': ['lasso2d', 'select2d']
    }
)

### 1. Кризис 2022 года остался невидимым для модели

Transformer не увидел кризис 2022 года. На графике он показал практически плоскую линию в момент, когда реальная волатильность резко взлетела. Это классическая проблема Transformer'ов в финансовых временных рядах:

Причина заключается в том, что Transformer полагается на self-attention (механизм внимания) для поиска паттернов во времени. В относительно спокойные периоды (2010-2021) модель выучила, что волатильность должна быть низкой и стабильной. Когда наступил кризис, модель не смогла адаптироваться - она продолжала предсказывать "спокойное" состояние, игнорируя резкие изменения.

### 2. Transformer требует гораздо больше данных для обучения

| Модель | Параметров | Train Loss | Val Loss |
|--------|-----------|------------|----------|
| LSTM | ~340,000 | 0.177 | 0.176 |
| Transformer | ~290,000 | 0.167 | 0.241 |

У Transformer ниже train loss, но выше val loss, чем у LSTM. Это скорее всего переобучение, несмотря на то, что параметров у него меньше, но это не отменяет проблемуу исключительного мехнизма фокусирвовки внимания. 

### 3. Временная структура данных не подходит для Attention

Финансовые временные ряды имеют локальную зависимость: волатильность сегодня сильно коррелирует с волатильностью вчера, но слабо - с тем, что было 3 месяца назад.

### 4. Эмпирическое подтверждение из метрик

| Метрика | Transformer vs LSTM |
|---------|---------------------|
| MSE | **в 1.8 раза хуже** (635 vs 350) |
| R² | **почти ноль** (0.13 vs 0.52) |

R² = 0.128 означает, что модель объясняет всего 13% дисперсии целевой переменной. Это близко к случайному предсказанию.

### 5. GARCH оказался лучшим по MSE и R²

Интересно, что классический GARCH обошёл обе нейросети по MSE и R². Это подтверждает тезис из статьи GINN:
> *GARCH по-прежнему остаётся одним из лучших моделей для волатильности, особенно в кризисные периоды* - III Results


## GINN (GARCH-Informed Neural Network)

Гибрид из статьи Xu et al. (2024, Carnegie Mellon).
Та же LSTM архитектура, но loss = λ·MSE(real, pred) + (1-λ)·MSE(garch, pred).

In [61]:
# GARCH rolling forecast для ВСЕХ окон (train + val + test)
# Это нужно чтобы GINN мог учиться у GARCH

print("Считаю GARCH предсказания для всех окон ")

garch_all_preds = []

for i in range(len(X_windows)):
    # Позиция в returns: WINDOW (90) + offset от rolling + i
    pos = ground_truth_var.dropna().index[WINDOW + i]
    train_data = returns.iloc[pos - 90 : pos]
    
    try:
        m = arch_model(train_data, vol='Garch', p=1, q=1, dist='Normal')
        r = m.fit(disp='off')
        f = r.forecast(horizon=1)
        garch_all_preds.append(f.variance.values[-1, 0])
    except:
        garch_all_preds.append(np.nan)
    
    if (i + 1) % 500 == 0:
        print(f"  {i+1}/{len(X_windows)} готово")

garch_all_preds = np.array(garch_all_preds)

nan_mask = np.isnan(garch_all_preds)
if nan_mask.sum() > 0:
    garch_all_preds[nan_mask] = np.nanmean(garch_all_preds)
    print(f"  Заполнено {nan_mask.sum()} NaN")

print(f"Готово: {len(garch_all_preds)} GARCH предсказаний")

Считаю GARCH предсказания для всех окон 
  500/3893 готово
  1000/3893 готово
  1500/3893 готово
  2000/3893 готово
  2500/3893 готово
  3000/3893 готово
  3500/3893 готово
Готово: 3893 GARCH предсказаний


In [62]:
# Нормализуем GARCH предсказания так же как y
garch_all_log = np.log1p(garch_all_preds)
garch_all_norm = (garch_all_log - train_mean) / train_std

# Split по тем же индексам
garch_train_norm = garch_all_norm[train_idx]
garch_val_norm = garch_all_norm[val_idx]
garch_test_norm = garch_all_norm[test_idx]

print(f"GARCH norm — train: {len(garch_train_norm)}, val: {len(garch_val_norm)}, test: {len(garch_test_norm)}")

GARCH norm — train: 2326, val: 505, test: 1062


In [63]:
# Dataset для GINN: X, y_real, y_garch

class GINNDataset(torch.utils.data.Dataset):
    def __init__(self, X, y_real, y_garch):
        self.X = torch.FloatTensor(X).unsqueeze(-1)  # (N, 90, 1)
        self.y_real = torch.FloatTensor(y_real)
        self.y_garch = torch.FloatTensor(y_garch)
    
    def __len__(self):
        return len(self.y_real)
    
    def __getitem__(self, idx):
        return self.X[idx], self.y_real[idx], self.y_garch[idx]

ginn_train = GINNDataset(X_train_norm, y_train_norm, garch_train_norm)
ginn_val = GINNDataset(X_val_norm, y_val_norm, garch_val_norm)
ginn_test = GINNDataset(X_test_norm, y_test_norm, garch_test_norm)

ginn_train_loader = DataLoader(ginn_train, batch_size=64, shuffle=True)
ginn_val_loader = DataLoader(ginn_val, batch_size=64, shuffle=False)
ginn_test_loader = DataLoader(ginn_test, batch_size=64, shuffle=False)

In [69]:
# === GINN с меньшей моделью ===
LAMBDA = 0.01

ginn_model = VolatilityLSTM(hidden_size=64, num_layers=2, dropout=0.1)
ginn_model.to(device)

ginn_optimizer = optim.AdamW(ginn_model.parameters(), lr=5e-4)
ginn_scheduler = optim.lr_scheduler.ReduceLROnPlateau(ginn_optimizer, mode='min', patience=30, factor=0.5)
mse_loss = nn.MSELoss()

best_val_loss = float('inf')
best_model_state = None
patience_counter = 0
PATIENCE = 50  # больше терпения

for epoch in range(500):
    ginn_model.train()
    train_losses = []
    
    for X_batch, y_real_batch, y_garch_batch in ginn_train_loader:
        X_batch = X_batch.to(device)
        y_real_batch = y_real_batch.to(device)
        y_garch_batch = y_garch_batch.to(device)
        
        ginn_optimizer.zero_grad()
        preds = ginn_model(X_batch)
        
        loss = LAMBDA * mse_loss(preds, y_real_batch) + (1 - LAMBDA) * mse_loss(preds, y_garch_batch)
        
        loss.backward()
        torch.nn.utils.clip_grad_norm_(ginn_model.parameters(), max_norm=1.0)
        ginn_optimizer.step()
        train_losses.append(loss.item())
    
    avg_train = np.mean(train_losses)
    
    ginn_model.eval()
    val_losses = []
    with torch.no_grad():
        for X_batch, y_real_batch, y_garch_batch in ginn_val_loader:
            preds = ginn_model(X_batch.to(device))
            loss = LAMBDA * mse_loss(preds, y_real_batch.to(device)) + (1 - LAMBDA) * mse_loss(preds, y_garch_batch.to(device))
            val_losses.append(loss.item())
    
    avg_val = np.mean(val_losses)
    ginn_scheduler.step(avg_val)
    
    if avg_val < best_val_loss:
        best_val_loss = avg_val
        best_model_state = {k: v.clone() for k, v in ginn_model.state_dict().items()}
        patience_counter = 0
    else:
        patience_counter += 1
    
    if (epoch + 1) % 10 == 0:
        print(f"Эпоха {epoch+1}/500 - Train: {avg_train:.6f}, Val: {avg_val:.6f}")
    
    if patience_counter >= PATIENCE:
        print(f"Ранняя остановка на эпохе {epoch+1}")
        break

ginn_model.load_state_dict(best_model_state)
print("GINN обучен!")

Эпоха 10/500 - Train: 0.094606, Val: 0.591941
Эпоха 20/500 - Train: 0.087305, Val: 0.684886
Эпоха 30/500 - Train: 0.082719, Val: 0.564859
Эпоха 40/500 - Train: 0.084458, Val: 0.461211
Эпоха 50/500 - Train: 0.076027, Val: 0.567929
Эпоха 60/500 - Train: 0.067585, Val: 0.599223
Эпоха 70/500 - Train: 0.065472, Val: 0.559618
Эпоха 80/500 - Train: 0.060241, Val: 0.570659
Эпоха 90/500 - Train: 0.061146, Val: 0.559239
Ранняя остановка на эпохе 90
GINN обучен!


In [70]:
ginn0_model = VolatilityLSTM(hidden_size=64, num_layers=2, dropout=0.1)
ginn0_model.to(device)

ginn0_optimizer = optim.AdamW(ginn0_model.parameters(), lr=5e-4)
ginn0_scheduler = optim.lr_scheduler.ReduceLROnPlateau(ginn0_optimizer, mode='min', patience=30, factor=0.5)

best_val_loss = float('inf')
best_model_state = None
patience_counter = 0

for epoch in range(500):
    ginn0_model.train()
    train_losses = []
    
    for X_batch, y_real_batch, y_garch_batch in ginn_train_loader:
        ginn0_optimizer.zero_grad()
        preds = ginn0_model(X_batch.to(device))
        loss = mse_loss(preds, y_garch_batch.to(device))
        loss.backward()
        torch.nn.utils.clip_grad_norm_(ginn0_model.parameters(), max_norm=1.0)
        ginn0_optimizer.step()
        train_losses.append(loss.item())
    
    avg_train = np.mean(train_losses)
    
    ginn0_model.eval()
    val_losses = []
    with torch.no_grad():
        for X_batch, y_real_batch, y_garch_batch in ginn_val_loader:
            preds = ginn0_model(X_batch.to(device))
            loss = mse_loss(preds, y_garch_batch.to(device))
            val_losses.append(loss.item())
    
    avg_val = np.mean(val_losses)
    ginn0_scheduler.step(avg_val)
    
    if avg_val < best_val_loss:
        best_val_loss = avg_val
        best_model_state = {k: v.clone() for k, v in ginn0_model.state_dict().items()}
        patience_counter = 0
    else:
        patience_counter += 1
    
    if (epoch + 1) % 10 == 0:
        print(f"Эпоха {epoch+1}/500 - Train: {avg_train:.6f}, Val: {avg_val:.6f}")
    
    if patience_counter >= PATIENCE:
        print(f"GINN-0: ранняя остановка на эпохе {epoch+1}")
        break

ginn0_model.load_state_dict(best_model_state)
print("GINN-0 обучен!")

Эпоха 10/500 - Train: 0.106659, Val: 0.851876
Эпоха 20/500 - Train: 0.079288, Val: 0.719370
Эпоха 30/500 - Train: 0.075928, Val: 0.583294
Эпоха 40/500 - Train: 0.070847, Val: 0.506476
Эпоха 50/500 - Train: 0.068547, Val: 0.560585
GINN-0: ранняя остановка на эпохе 53
GINN-0 обучен!


In [71]:
# Оценка GINN и GINN-0
def evaluate_model(model, loader, name):
    model.eval()
    preds_list = []
    with torch.no_grad():
        for batch in loader:
            X_batch = batch[0].to(device)
            preds = model(X_batch).cpu().numpy()
            preds_list.extend(preds)
    preds_arr = np.array(preds_list)
    preds_denormed = np.expm1(preds_arr * train_std + train_mean)
    
    m_mse = mean_squared_error(targets_denorm, preds_denormed)
    m_mae = mean_absolute_error(targets_denorm, preds_denormed)
    m_r2 = r2_score(targets_denorm, preds_denormed)
    return preds_denormed, m_mse, m_mae, m_r2

ginn_preds_denorm, ginn_mse, ginn_mae, ginn_r2 = evaluate_model(ginn_model, ginn_test_loader, "GINN")
ginn0_preds_denorm, ginn0_mse, ginn0_mae, ginn0_r2 = evaluate_model(ginn0_model, ginn_test_loader, "GINN-0")

print(f"{'Модель':<15} {'MSE':>10} {'MAE':>10} {'R²':>10}")
print(f"{'GARCH':<15} {garch_mse:>10.4f} {garch_mae:>10.4f} {garch_r2:>10.4f}")
print(f"{'LSTM':<15} {mse:>10.4f} {mae:>10.4f} {r2:>10.4f}")
print(f"{'Transformer':<15} {trans_mse:>10.4f} {trans_mae:>10.4f} {trans_r2:>10.4f}")
print(f"{'GINN (λ=0.01)':<15} {ginn_mse:>10.4f} {ginn_mae:>10.4f} {ginn_r2:>10.4f}")
print(f"{'GINN-0 (λ=0)':<15} {ginn0_mse:>10.4f} {ginn0_mae:>10.4f} {ginn0_r2:>10.4f}")

Модель                 MSE        MAE         R²
GARCH             261.2266     2.3915     0.6418
LSTM              350.9200     1.9464     0.5188
Transformer       635.6202     2.4518     0.1283
GINN (λ=0.01)     692.5250     3.2562     0.0503
GINN-0 (λ=0)      689.5466     3.3046     0.0544


In [72]:
fig = go.Figure()
fig.add_trace(go.Scatter(x=dates_test, y=targets_denorm, name='Реальность', opacity=0.4))
fig.add_trace(go.Scatter(x=dates_test, y=garch_preds_raw, name='GARCH', opacity=0.7))
fig.add_trace(go.Scatter(x=dates_test, y=preds_denorm, name='LSTM', opacity=0.7))
fig.add_trace(go.Scatter(x=dates_test, y=ginn_preds_denorm, name='GINN', opacity=0.7))
fig.add_trace(go.Scatter(x=dates_test, y=ginn0_preds_denorm, name='GINN-0', opacity=0.7))
fig.update_layout(title='Все модели vs реальность', xaxis_title='Дата', yaxis_title='σ²')
fig.show()